Implementation of xp command

ahmedhozny opened the pull request:
## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
Implement XP command as in vanilla 

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
Added new class XPCommand which has the command basics
### Behavioural changes
<!-- Any change in how the server behaves, or its performance? -->
N/A
## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->
N/A

Requires translations:

| Name | Value in eng.ini |
| `commands.xp.success.other` | `Gave {%0} to {%1)` |
| `commands.xp.success.self` | `Gave himself {%0}` |
| `pocketmine.command.xp.usage` | `/xp <experience[L]> <player>` |

-->

## Tests
I've tried the command with different situations (Player offline or not found, L letter adds level, Op can give to himself, etc..)

<!--
Details should be provided of tests done. Simply saying "tested" or equivalent is not acceptable.

Attach scripts or actions to test this pull request, as well as the result
-->

ahmedhozny committed:
Add files via upload

ahmedhozny committed:
Add files via upload

ahmedhozny committed:
Full and final implementation of XP command

ahmedhozny committed:
Ok this is the final one

ahmedhozny committed:
Fixes and improvements

ahmedhozny committed:
Add files via upload

ahmedhozny committed:
Improvments (again)

ahmedhozny committed:
Update SimpleCommandMap.php

ahmedhozny committed:
Update DefaultPermissions.php

ahmedhozny committed:
Update XPCommand.php

ahmedhozny committed:
Update XPCommand.php

ahmedhozny committed:
Added self feature

lukeeey commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+                        }
+                        
+                }elseif(!is_numeric($args[0])){
+                        $sender->sendMessage(TextFormat::RED. "Please write in numbers");
```
"Please write in numbers" doesn't sound like very good English.

dktapps replied:
This should be a generic command syntax error.

lukeeey commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+use pocketmine\lang\TranslationContainer;
+use pocketmine\utils\TextFormat;
+
+class XPCommand extends VanillaCommand{
```
I think this should be `XpCommand` instead of `XPCommand`

lukeeey commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+		        "%pocketmine.command.xp.description",
+		        "%pocketmine.command.xp.usage"
+		);
+	$this->setPermission("pocketmine.command.xp");
```
Indentation throughout entire file

dktapps commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+	        }
+                
+                if(count($args) < 1){
+	                throw new InvalidCommandSyntaxException();
```
Throwing an `InvalidCommandSyntaxException` makes the usage message be displayed. That's its reason for existence.

dktapps commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+                        return true;
+                } 
+                if(isset($args[1])){
+                        $player->addXp(intval($args[0]));
```
use a typecast, not `intval()`.

dktapps commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+                                Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.xp.success.self",[$xp]));
+                                return true;
+                        }
+                        
```
The returns in both of the above blocks are redundant and make the `else` useless. They should be placed outside of it. Don't litter the function with lots of return points.

dktapps commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+                                        $xp,
+                                        $player->getName()
+                                ]));
+                        return true;
```
... formatting?

dktapps commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+                }
+                
+                if(strtolower(preg_replace("/[^a-zA-Z]/", "",$args[0])) == "l"){
+                        $xp = intval(preg_replace("/[^0-9]+/", "", $args[0]), 10);
```
This regex doesn't make sense. If I read correctly `5l0` will be interpreted as 50 levels. And again, use a typecast, not intval().

dktapps commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+                        }
+                }
+                
+                if(strtolower(preg_replace("/[^a-zA-Z]/", "",$args[0])) == "l"){
```
this should use strict equality check.

dktapps commented on src/pocketmine/command/defaults/XPCommand.php:
```diff
@@ -0,0 +1,90 @@
+                if(isset($args[1])){
+                        $player->addXp(intval($args[0]));
+                        Command::broadcastCommandMessage($sender, new TranslationContainer("commands.xp.success", [
+                                intval($args[0]),
```
You have a tendecy to repeatedly call functions applying operations to the same thing. Apply it once, store it in a variable and use the variable.

ahmedhozny committed:
Fixed a lot of issues and made some imporvements

ahmedhozny committed:
oops forgot to delete this one

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,98 @@
+                                ]));
+                        }else{
+                                if(!($sender instanceof Player)){
+                                        throw new InvalidCommandSyntaxException;
```
missing `()`

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,98 @@
+                            if(!($sender instanceof Player)){
+                                throw new InvalidCommandSyntaxException;
+                            }
+                                $sender->getServer()->getPlayer($sender->getName())->addXp($xp);
```
... what in god's name is this?

dschwartz783 replied:
This code actually made me laugh. xD

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,98 @@
+                        return true;
+                }
+                
+                if(is_numeric($args[0])){
```
This should be using `if..elseif`, not abusing return statements.

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,98 @@
+                                ]));
+                        }else{
+                            if(!($sender instanceof Player)){
+                                throw new InvalidCommandSyntaxException;
```
nit: missing `()`

lukeeey replied:
what does `nit` mean?

killer549 replied:
## @dktapps Yeah and would you please specify the issue here

dktapps replied:
are you capable of reading?

> missing `()`

dktapps replied:
@lukeeey it means small issue in this context.

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,98 @@
+                }
+                
+                if(strtolower(substr($args[0], -1)) == "l"){
+                        $lxp = $this->getInteger($sender, substr_replace($args[0], "", -1));
```
... why not simply `substr($args[0], 0, -1)`

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,98 @@
+                        }
+                }
+                
+                if(strtolower(substr($args[0], -1)) == "l"){
```
again, use strict comparison, not weak.

dktapps reviewed (changes requested):
I hope you don't expect this to be merged with such abominable formatting...

ahmedhozny committed:
Fixed some stupid issues

ahmedhozny committed:
Formate and style #1

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -58,38 +58,30 @@ public function execute(CommandSender $sender, string $commandLabel, array $args
                 
-                if(strtolower(substr($args[0], -1)) == "l"){
-                        $lxp = $this->getInteger($sender, substr_replace($args[0], "", -1));
+                if(strcasecmp(substr($args[0], -1), "l") == 0){
```
Why change this?

killer549 replied:
I thought it would be a better comparison

dktapps replied:
it's a completely unnecessary alteration.

ahmedhozny committed:
Formatted

ahmedhozny committed:
.

dschwartz783 committed:
Complete formatting fix

ahmedhozny committed:
Merge pull request #1 from dschwartz783/master

Complete formatting fix

lukeeey commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -58,38 +58,30 @@ public function execute(CommandSender $sender, string $commandLabel, array $args
+                                throw new InvalidCommandSyntaxException();
                             }
-                                $sender->getServer()->getPlayer($sender->getName())->addXp($xp);
-                                Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.xp.success.self",[$xp]));
```
Umm didnt someone say to keep the % there? It means it will be translated client side.

lukeeey commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -58,38 +58,30 @@ public function execute(CommandSender $sender, string $commandLabel, array $args
-                                        $player->getName()
-                                ]));
+                }elseif(is_numeric($args[0])){
+                        $xp = $this->getInteger($sender, $args[0]);
```
Why not use a type cast?
```php
$xp = (int) $args[0];
```

killer549 commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,90 @@
+
+		if(strtolower(substr($args[0], -1)) == "l"){
+			$lxp = $this->getInteger($sender, substr($args[0], 0, -1));
+			if(isset($args[1])){
```
@dktapps Sorry for that but is that even possible to do without changing about quarter of the code ?😕

dschwartz783 replied:
If $args[1] is set, then earlier in your code you got a Player object to work with. Since you're not actually using $args[1] in the following code, but rather the Player object you created earlier, you should just check for that. Although, you could probably refactor this code a bit so that this entire block (as well as the other block where you check for isset($args[1]) appears in the same block where you initially created the Player object. It would reduce code duplication a bit.

ahmedhozny committed:
Add files via upload

dschwartz783 commented:
You destroyed the formatting again. Tabs, not spaces.

killer549 commented:
Not sure why am I using Netbeans

dschwartz783 commented:
Use PHPStorm.

dktapps commented:
please read... https://github.com/pmmp/PocketMine-MP/blob/master/CONTRIBUTING.md#code-contributions

ahmedhozny committed:
Format improvements...

dschwartz783 commented:
My point about using PHPStorm which I suppose I should have mentioned, is that PHPStorm will automatically handle some formatting issues, by dynamically switching to either tabs or spaces for indentation depending on the file. I don't believe NetBeans does this. The suggestion was not off-topic.

ahmedhozny committed:
Better format (I'm tired of this)

ahmedhozny committed:
oops wrong path

ahmedhozny committed:
Add files via upload

Muqsit commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,92 @@
+				$player->addXp($xp);
+				Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.xp.success.other", [$xp, $player->getName()]));
+			}else{
+				$player->addXp($xp);
```
you can move this just after line 80 and remove this and line 82... right?
```php
$player->addXp($xp);
if($player !== $sender){
    Command::broadc...
}else{
    Command::broadc...
}
```

killer549 replied:
Nice idea..
Thanks for your suggestion. 😃

ahmedhozny committed:
Improvements

robske110 commented:
<img width="962" alt="image" src="https://user-images.githubusercontent.com/16020100/38954344-a9d6fa38-4351-11e8-8c6c-a3e88618b305.png">
Still mixed tabs&spaces here

ahmedhozny committed:
unnoticed formatting error

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,90 @@
+				Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.xp.success.other", [$xp, $player->getName()]));
+			}else{
+				Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.xp.success.self",[$xp]));
+			}
```
code copy-paste - also, these messages should not be the same for both levels and XP points.

killer549 replied:
Is it better to remove `self` and `other` and just provide on message like this `"%commands.xp.success`?

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,90 @@
+				throw new InvalidCommandSyntaxException();			
+			}
+			
+			$lxp = (int) substr($args[0], 0, -1);
```
the string after removing the `l` might not be numeric, in which case this will produce undefined behaviour.

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,90 @@
+			if($player !== $sender){
+				Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.xp.success.other", [$lxp, $player->getName()]));
+			}else{
+				Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.xp.success.self",[$lxp]));
```
formatting issue here (and in the copy pasta block below)

dktapps commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -0,0 +1,90 @@
+		if(strtolower(substr($args[0], -1)) === "l"){
+			if(!($player instanceof Player)){
+				throw new InvalidCommandSyntaxException();			
+			}
```
why duplicate this check in multiple branches? this should be outside of here, where you actually get the sender in the first place.

Additionally I don't think a command-syntax error is accurate in this case - it should be something like `Please specify a player` as per things like /teleport.

lukeeey replied:
Why not allow InvalidCommandSyntaxException to have a string parameter?

dktapps replied:
That is outside the scope of this PR - and regardless, this isn't a command syntax error, it's console not being permitted to use a player-specific overload.

ahmedhozny committed:
fixes and improvements

ahmedhozny committed:
Fixed a bug

Fixed a bug appears when a player is exposed to have negative experience or levels

killer549 commented on src/pocketmine/command/defaults/XpCommand.php:
```diff
@@ -68,13 +68,34 @@ public function execute(CommandSender $sender, string $commandLabel, array $args
+						return true;
+					}
+					
+					$player->subtractXpLevels(-$level);
```
Not sure if using subtract here (and in the block below) is a great idea but I just didn't want to put `$player->addXpLevels($level);` at last after everything has done.
